File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/no-render-return-value.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { Rule } from "../../utils/rule.js";

const MESSAGE = "Your app breaks in React 19 because `ReactDOM.render` returns nothing there.";

const isReactDomRenderCall = (node: EsTreeNodeOfType<"CallExpression">): boolean => {
  if (!isNodeOfType(node.callee, "MemberExpression")) return false;
  if (!isNodeOfType(node.callee.object, "Identifier")) return false;
  if (node.callee.object.name !== "ReactDOM") return false;
  if (!isNodeOfType(node.callee.property, "Identifier")) return false;
  return node.callee.property.name === "render";
};

const isUsedAsReturnValue = (parent: EsTreeNode | null | undefined): boolean => {
  if (!parent) return false;
  if (
    isNodeOfType(parent, "VariableDeclarator") ||
    isNodeOfType(parent, "Property") ||
    isNodeOfType(parent, "ReturnStatement") ||
    isNodeOfType(parent, "AssignmentExpression")
  ) {
    return true;
  }
  // Expression-bodied arrow function: `() => ReactDOM.render(...)`
  if (
    isNodeOfType(parent, "ArrowFunctionExpression") &&
    parent.body &&
    !isNodeOfType(parent.body, "BlockStatement")
  ) {
    return true;
  }
  return false;
};

// Port of `oxc_linter::rules::react::no_render_return_value`. Reports when
// the return value of `ReactDOM.render(...)` is captured into a variable,
// returned from a function, assigned, used as an object-property value,
// or implicitly returned from an expression-bodied arrow.
export const noRenderReturnValue = defineRule<Rule>({
  id: "no-render-return-value",
  title: "Using ReactDOM.render return value",
  severity: "warn",
  recommendation:
    "Don't use `ReactDOM.render`'s return value. It's legacy and was removed in React 19.",
  create: (context) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      if (!isReactDomRenderCall(node)) return;
      if (!isUsedAsReturnValue(node.parent)) return;
      context.report({ node: node.callee, message: MESSAGE });
    },
  }),
});

File: native/oxlint/rules/no-render-return-value.rs
use oxc_ast::{AstKind, ast::Expression};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{AstNode, context::LintContext, rule::Rule};

const MESSAGE: &str = "Your app breaks in React 19 because `ReactDOM.render` returns nothing there.";

#[derive(Debug, Default, Clone)]
pub struct NoRenderReturnValue;

declare_oxc_lint!(
    /// Disallow using the return value of `ReactDOM.render()`.
    NoRenderReturnValue,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow using the ReactDOM.render return value.",
);

impl Rule for NoRenderReturnValue {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(call_expression) = node.kind() else {
            return;
        };
        let Expression::StaticMemberExpression(member_expression) =
            call_expression.callee.get_inner_expression()
        else {
            return;
        };
        let Expression::Identifier(receiver) = member_expression.object.get_inner_expression()
        else {
            return;
        };
        if receiver.name != "ReactDOM" || member_expression.property.name != "render" {
            return;
        }
        let parent = ctx.nodes().parent_node(node.id());
        if !matches!(
            parent.kind(),
            AstKind::VariableDeclarator(_)
                | AstKind::ObjectProperty(_)
                | AstKind::ReturnStatement(_)
                | AstKind::AssignmentExpression(_)
                | AstKind::ArrowFunctionExpression(_)
        ) {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(call_expression.callee.span()));
    }
}
